File: Sources/TokenTrackerCore/CredentialSource.swift
import Foundation
#if canImport(Security)
import Security
#endif

/// Where the Claude access token actually came from, and when it was last written.
///
/// Diagnostics used to report only whether `~/.claude/.credentials.json` exists,
/// which reads `false` on a normal macOS install — the token lives in the keychain
/// and the file is a fallback. That hid the question that matters when usage stops
/// loading: *is anything still refreshing this token?*
///
/// The access token lives ~8h and only the Claude Code CLI (or its daemon, which
/// exits when idle) writes this store; the desktop app refreshes its own auth
/// elsewhere and leaves the keychain item untouched. So a modification date hours
/// or days old is the signal that the store has been abandoned, and the whole
/// answer to "why does it keep expiring".
///
/// Reads attributes only — never the secret.
public struct ClaudeCredentialSource: Sendable, Equatable {
    public enum Kind: String, Sendable {
        case keychain
        case credentialsFile
        case none
    }

    public let kind: Kind
    public let lastWrittenAt: Date?

    public init(kind: Kind, lastWrittenAt: Date?) {
        self.kind = kind
        self.lastWrittenAt = lastWrittenAt
    }

    public static func detect(credentialsFileURL: URL) -> ClaudeCredentialSource {
        detect(credentialsFileURL: credentialsFileURL, keychainModifiedAt: keychainModificationDate())
    }

    /// Seam for tests: the keychain lookup is the one part that cannot be arranged
    /// on a build machine, and the precedence between the two stores is exactly what
    /// the old "does the file exist" line got wrong.
    public static func detect(credentialsFileURL: URL, keychainModifiedAt: Date?) -> ClaudeCredentialSource {
        if let keychainModifiedAt {
            return ClaudeCredentialSource(kind: .keychain, lastWrittenAt: keychainModifiedAt)
        }
        if let attributes = try? FileManager.default.attributesOfItem(atPath: credentialsFileURL.path) {
            return ClaudeCredentialSource(
                kind: .credentialsFile,
                lastWrittenAt: attributes[.modificationDate] as? Date
            )
        }
        return ClaudeCredentialSource(kind: .none, lastWrittenAt: nil)
    }

    /// How stale the store is, which is what tells a lapsed token apart from a
    /// merely unlucky one: past roughly the 8h access-token life, nothing is
    /// refreshing it any more.
    public func age(now: Date = Date()) -> TimeInterval? {
        lastWrittenAt.map { now.timeIntervalSince($0) }
    }

    private static func keychainModificationDate() -> Date? {
        #if canImport(Security)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: "Claude Code-credentials",
            kSecReturnAttributes as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var item: CFTypeRef?
        guard
            SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
            let attributes = item as? [String: Any]
        else {
            return nil
        }
        return attributes[kSecAttrModificationDate as String] as? Date
        #else
        return nil
        #endif
    }
}

/// The line the menu was missing when Claude usage stops loading: nothing has
/// written the credential store for longer than a token lives, so the fix is not
/// to wait but to run the CLI once.
///
/// The plain "sign in again" recovery is complete advice for a token that merely
/// lapsed. It is not the whole story for a store that has been abandoned — a user
/// who signs in through the desktop app has signed in, and the store still will
/// not move. Saying how long it has been unwritten is what tells the two apart.
public enum CredentialStoreAdvice {
    /// A Claude access token lives ~8h. Past that, an expired token is not bad
    /// luck: whatever used to refresh the store has stopped.
    public static let staleAfter: TimeInterval = 8 * 3600

    /// `nil` unless this is Claude, the failure is a lapsed token, and the store
    /// is old enough that nothing can still be refreshing it.
    public static func staleLine(
        provider: Provider,
        issue: UsageIssue,
        source: ClaudeCredentialSource,
        now: Date = Date(),
        localizer: Localizer = Localizer(language: .english)
    ) -> String? {
        guard provider == .claude else { return nil }
        // Read past the visible kind to the error underneath it: once the last good
        // reading is being carried across a lapsed token the status reads "using
        // cached data", and the expiry survives only in the error it carries.
        guard
            let error = issue.technicalDetail,
            UsageIssueFormatter.kind(forError: error) == .expiredCredentials,
            let age = source.age(now: now),
            age >= staleAfter
        else {
            return nil
        }
        return localizer.text(.credentialStale)
            .replacingOccurrences(of: "{age}", with: UsageForecaster.durationText(age))
    }
}

File: windows/TokenTracker.Windows.Core/CredentialSource.cs
namespace TokenTracker.Windows.Core;

public enum ClaudeCredentialStore
{
    Keychain,
    CredentialsFile,
    None
}

/// <summary>
/// Where the Claude access token actually came from, and when it was last written.
///
/// Diagnostics used to report only whether the credentials file exists, which says
/// nothing about the question that matters when usage stops loading: is anything
/// still refreshing this token? The access token lives ~8h and only the Claude Code
/// CLI writes the store, so a modification date hours or days old is the signal that
/// the store has been abandoned.
///
/// Reads attributes only — never the secret. Windows has no keychain, so the file is
/// always the live store here; the keychain seam exists so the precedence rule and
/// its tests stay identical to the macOS port.
/// </summary>
public sealed record ClaudeCredentialSource(ClaudeCredentialStore Kind, DateTimeOffset? LastWrittenAt)
{
    public static ClaudeCredentialSource Detect(string credentialsFilePath, DateTimeOffset? keychainModifiedAt = null)
    {
        if (keychainModifiedAt is { } modified)
        {
            return new ClaudeCredentialSource(ClaudeCredentialStore.Keychain, modified);
        }

        if (File.Exists(credentialsFilePath))
        {
            return new ClaudeCredentialSource(
                ClaudeCredentialStore.CredentialsFile,
                new DateTimeOffset(File.GetLastWriteTimeUtc(credentialsFilePath), TimeSpan.Zero));
        }

        return new ClaudeCredentialSource(ClaudeCredentialStore.None, null);
    }

    /// <summary>
    /// How stale the store is, which is what tells a lapsed token apart from a merely
    /// unlucky one: past roughly the 8h access-token life, nothing is refreshing it.
    /// </summary>
    public TimeSpan? Age(DateTimeOffset now) =>
        LastWrittenAt is { } written ? now - written : null;
}

/// <summary>
/// The line the menu was missing when Claude usage stops loading: nothing has written
/// the credential store for longer than a token lives, so the fix is not to wait but
/// to run the CLI once.
///
/// The plain "sign in again" recovery is complete advice for a token that merely
/// lapsed. It is not the whole story for a store that has been abandoned — a user who
/// signs in through the desktop app has signed in, and the store still will not move.
/// Saying how long it has been unwritten is what tells the two apart.
/// </summary>
public static class CredentialStoreAdvice
{
    /// <summary>
    /// A Claude access token lives ~8h. Past that, an expired token is not bad luck:
    /// whatever used to refresh the store has stopped.
    /// </summary>
    public static readonly TimeSpan StaleAfter = TimeSpan.FromHours(8);

    /// <summary>
    /// Null unless this is Claude, the failure is a lapsed token, and the store is old
    /// enough that nothing can still be refreshing it.
    /// </summary>
    public static string? StaleLine(
        Provider provider,
        UsageIssue issue,
        ClaudeCredentialSource source,
        DateTimeOffset now,
        Localizer? localizer = null)
    {
        if (provider != Provider.Claude || string.IsNullOrWhiteSpace(issue.TechnicalDetail))
        {
            return null;
        }

        // Read past the visible kind to the error underneath it: once the last good
        // reading is being carried across a lapsed token the status reads "using cached
        // data", and the expiry survives only in the error it carries.
        if (UsageIssueFormatter.Kind(issue.TechnicalDetail) != UsageIssueKind.ExpiredCredentials)
        {
            return null;
        }

        if (source.Age(now) is not { } age || age < StaleAfter)
        {
            return null;
        }

        localizer ??= new Localizer(AppLanguage.English);
        return localizer
            .Text(L10nKey.CredentialStale)
            .Replace("{age}", UsageForecaster.DurationText(age.TotalSeconds), StringComparison.Ordinal);
    }
}
